File: internal/core/entity/deck.go
package entity

import (
	"fmt"
	"time"
)

const (
	// DeckSize デッキの枚数（固定）
	DeckSize = 40
	// MaxCardCopies 同じカードの最大枚数
	MaxCardCopies = 3
	// MaxDeckNameLength デッキ名の最大長
	MaxDeckNameLength = 100
	// MaxDeckDescriptionLength デッキ説明の最大長
	MaxDeckDescriptionLength = 500
)

// Deck はデッキを表すドメインエンティティ
type Deck struct {
	ID          string
	Name        string
	Description string
	CardIDs     []string
	UserID      string
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// NewDeck は新しいDeckエンティティを作成
func NewDeck(id, name, description, userID string, cardIDs []string) (*Deck, error) {
	now := time.Now()
	deck := &Deck{
		ID:          id,
		Name:        name,
		Description: description,
		CardIDs:     cardIDs,
		UserID:      userID,
		CreatedAt:   now,
		UpdatedAt:   now,
	}

	if err := deck.Validate(); err != nil {
		return nil, err
	}

	return deck, nil
}

// Validate はデッキの妥当性を検証
func (d *Deck) Validate() error {
	// 名前検証
	if d.Name == "" {
		return NewErrInvalidDeck("name", "デッキ名は必須です")
	}
	if len(d.Name) > MaxDeckNameLength {
		return NewErrInvalidDeck("name", fmt.Sprintf("デッキ名は%d文字以内である必要があります", MaxDeckNameLength))
	}

	// 説明検証
	if len(d.Description) > MaxDeckDescriptionLength {
		return NewErrInvalidDeck("description", fmt.Sprintf("デッキ説明は%d文字以内である必要があります", MaxDeckDescriptionLength))
	}

	// ユーザーID検証
	if d.UserID == "" {
		return NewErrInvalidDeck("user_id", "ユーザーIDは必須です")
	}

	// カード数検証
	if len(d.CardIDs) == 0 {
		return NewErrInvalidDeck("cards", "デッキにカードが含まれていません")
	}

	// デッキの枚数は40枚固定
	if len(d.CardIDs) != DeckSize {
		return NewErrInvalidDeck("cards", fmt.Sprintf("デッキはちょうど%d枚である必要があります", DeckSize))
	}

	// 同じカードの枚数チェック
	cardCount := make(map[string]int)
	for _, cardID := range d.CardIDs {
		// 空のカードIDをチェック
		if cardID == "" {
			return NewErrInvalidDeck("cards", "無効なカードIDが含まれています")
		}

		cardCount[cardID]++
		if cardCount[cardID] > MaxCardCopies {
			return NewErrInvalidDeck("cards", fmt.Sprintf("同じカードは%d枚までしか入れられません", MaxCardCopies))
		}
	}

	return nil
}

// UpdateCards はカードリストを更新
func (d *Deck) UpdateCards(cardIDs []string) error {
	oldCardIDs := d.CardIDs
	d.CardIDs = cardIDs
	d.UpdatedAt = time.Now()

	if err := d.Validate(); err != nil {
		d.CardIDs = oldCardIDs // ロールバック
		return err
	}

	return nil
}

// UpdateInfo はデッキ名と説明を更新
func (d *Deck) UpdateInfo(name, description string) error {
	if name == "" {
		return NewErrInvalidDeck("name", "デッキ名は必須です")
	}
	if len(name) > MaxDeckNameLength {
		return NewErrInvalidDeck("name", fmt.Sprintf("デッキ名は%d文字以内である必要があります", MaxDeckNameLength))
	}
	if len(description) > MaxDeckDescriptionLength {
		return NewErrInvalidDeck("description", fmt.Sprintf("デッキ説明は%d文字以内である必要があります", MaxDeckDescriptionLength))
	}

	d.Name = name
	d.Description = description
	d.UpdatedAt = time.Now()
	return nil
}

File: frontend/src/domain/models/Deck.ts
/**
 * ドメイン層: デッキモデル
 */

export interface Deck {
  id: string
  name: string
  description: string
  cardIds: string[]
  userId: string
  createdAt: Date
  updatedAt: Date
}

export const MIN_DECK_SIZE = 30
export const MAX_DECK_SIZE = 30
export const MAX_CARD_COPIES = 3

export function validateDeck(deck: Partial<Deck>): string[] {
  const errors: string[] = []

  if (!deck.name || deck.name.trim() === '') {
    errors.push('デッキ名は必須です')
  }

  if (!deck.cardIds || deck.cardIds.length === 0) {
    errors.push('デッキには最低1枚のカードが必要です')
  }

  if (deck.cardIds && deck.cardIds.length < MIN_DECK_SIZE) {
    errors.push(`デッキには${MIN_DECK_SIZE}枚のカードが必要です`)
  }

  if (deck.cardIds && deck.cardIds.length > MAX_DECK_SIZE) {
    errors.push(`デッキには${MAX_DECK_SIZE}枚までしか入れられません`)
  }

  // 同じカードのコピー数チェック
  if (deck.cardIds) {
    const cardCounts = new Map<string, number>()
    for (const cardId of deck.cardIds) {
      cardCounts.set(cardId, (cardCounts.get(cardId) || 0) + 1)
    }

    for (const [cardId, count] of cardCounts.entries()) {
      if (count > MAX_CARD_COPIES) {
        errors.push(
          `カード ${cardId} は${MAX_CARD_COPIES}枚までしか入れられません`,
        )
      }
    }
  }

  return errors
}

export function isValidDeck(deck: Partial<Deck>): boolean {
  return validateDeck(deck).length === 0
}
